compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -510,18 +593,53 @@ abstract class GenJSCode extends plugins.PluginComponent
      *  The fields are initialized with the zero of their types.
      */
     def genClassFields(cd: ClassDef): List[js.FieldDef] = {
+      val classSym = cd.symbol
+      assert(currentClassSym.get == classSym,
+          "genClassFields called with a ClassDef other than the current one")
+
       // Non-method term members are fields, except for module members.
       (for {
-        f <- currentClassSym.info.decls
+        f <- classSym.info.decls
         if !f.isMethod && f.isTerm && !f.isModule
       } yield {
         implicit val pos = f.pos
+
         val mutable =
           suspectFieldMutable(f) || unexpectedMutatedFields.contains(f)
-        js.FieldDef(encodeFieldSym(f), toIRType(f.tpe), mutable)
+        val name =
+          if (isExposed(f)) js.StringLiteral(jsNameOf(f))
+          else encodeFieldSym(f)
+
+        val irTpe = {
+          if (f.tpe.typeSymbol != CharClass || !isScalaJSDefinedJSClass(classSym))
+            toIRType(f.tpe)
+          else
+            jstpe.ClassType(ir.Definitions.BoxedCharacterClass)
+        }
```

Comment:
Scalac has the same issue:

```scala
scala> class Foo(val x: Int) extends AnyVal
defined class Foo

scala> class A[T] { var x: T = _ }
defined class A

scala> new A[Foo]
res1: A[Foo] = A@39d186

scala> res1.x.x
java.lang.NullPointerException
  ... 43 elided
```

But currently, we are doing it wrong differently (value class fields never work, right? Or did I get this wrong?). We should probably replicate Scalac's behavior...